nemo_skills/code_execution/sandbox.py

```diff
@@ -372,42 +374,76 @@ def update_fn(map_to_future, line_dict, answer_format="natural_language"):
+                    if answer_format == "natural_language":
+                        if not use_predicted_answer_key:
                             line_dict["predicted_answer"] = extract_answer(
                                 line_dict["generation"],
                                 extract_from_boxed=extract_from_boxed,
                                 extract_regex=extract_regex,
                             )
-                        elif answer_format == "lean":
-                            line_dict["predicted_answer"] = (
-                                line_dict["header"]
-                                + line_dict["formal_statement"]
-                                + (
-                                    line_dict["generation"][:-3]
-                                    if line_dict["generation"].endswith("```")
-                                    else line_dict["generation"]
+                        else:
+                            if "predicted_answer" not in line_dict:
+                                raise ValueError(
+                                    "predicted_answer key not found in the line_dict. "
+                                    "Set use_predicted_answer_key=False to re-extract"
                                 )
+                    # TODO manually removing <｜begin▁of▁sentence｜> for trtllm
+                    elif answer_format == "lean":
+                        if not use_predicted_proof_key:
+                            # line_dict["predicted_proof"] = line_dict["header"] + line_dict["formal_statement"] + line_dict["generation"][:-3] if line_dict["generation"].endswith("```") else line_dict["generation"]
+                            ## TODO temp changes, need to add header to autoformalization step
+                            header = "import Mathlib\n\nopen Complex Filter Function Metric Finset\nopen scoped BigOperators Topology\n\n"
+
+                            # Extract generation, removing trailing ``` if present
+                            if line_dict["generation"].endswith("```"):
+                                generation = line_dict["generation"][:-3]
+                            else:
+                                generation = line_dict["generation"]
+
+                            generation = re.sub(r"^<｜begin▁of▁sentence｜>", "", generation)
+
+                            # Always include header and formal_statement
+                            line_dict["predicted_proof"] = header + line_dict["formal_statement"] + generation
+
+                        else:
+                            if "predicted_proof" not in line_dict:
+                                raise ValueError(
+                                    "predicted_proof key not found in the line_dict. "
+                                    "Set use_predicted_proof_key=False to re-combine"
+                                )
+                    elif answer_format == "lean-stat":
```

Comment:
what's the difference between this one and above one?

Reply:
2 main differences
1) We manually add header to the check, since it is not in test.jsonl
2) we add `sorry` to the end of generation, since we have `sorry` as a stop word.

Reply:
can we unify this by using that as default header? With something like `line_dict.get('header', ...)`? And for sorry, can you try to remove it from few-shots? If you have \n\n\n\n\n\n, "sorry" stop phrase shouldn't be needed, if the model follows instructions, I guess